src/links.jl

```diff
@@ -0,0 +1,139 @@
+"""
+    AbstractLink
+
+Abstract type defining maps from R^n -> X.
+They can be applied by calling `link(x)`
+
+A series of definitions are given in http://web.pdx.edu/~newsomj/mvclass/ho_link.pdf
+"""
+abstract type AbstractLink end
+
+(f::AbstractLink)(x) = apply(f, x)
```

Comment:
Is there a particular reason for defining `apply`, rather than just implementing `(f::ConcreteLink)(x)` directly?

Reply:
Not that this is complicated here, but isn't it complicated to give ChainRules for functor functions?

Reply:
It shouldn't be. I mean, all `rrule`s are of the form
```julia
rrule(config, f, args...)
```
or
```julia
rrule(f, args...)
```
and I think the testing infrastructure now handles this properly (Lyndon and Miha at Invenia have been working full time on ChainRules for the last couple of months -- they've made loads of progress and it's really very exciting).

Reply:
Ok I will get read of `apply`